### PR TITLE
Add type annotations and mypy checks for `cuda.parallel` 

### PR DESCRIPTION
**Pull request by shwina**
## Description

This PR adds type annotations for `cuda.parallel`, as well as a MyPy (+ pre-commit) config for static type checking.

This PR should be merged after https://github.com/NVIDIA/cccl/pull/3177/

## Checklist

- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cuda_parallel/cuda/parallel/experimental/_cccl.py [hunk 1]
```diff
@@ -6,7 +6,12 @@
 import numba
 import functools
 import ctypes
+import numpy as np
 from numba import types, cuda
+from typing import TYPE_CHECKING
+
+if TYPE_CHECKING:
+    from .iterators._iterators import IteratorBase
```
Comment (rwgk) [inline python/cuda_parallel/cuda/parallel/experimental/_cccl.py:14]: Do we really need to make this conditional on `TYPE_CHECKING`?

We're in a private file here. Simpler/straightforward is better?

File: python/cuda_parallel/cuda/parallel/experimental/_cccl.py [hunk 2]
```diff
@@ -197,14 +201,14 @@ def type_enum_as_name(enum_value):
     )[enum_value]
 
 
-def to_cccl_iter(array_or_iterator):
+def to_cccl_iter(array_or_iterator) -> Iterator:
     from cuda.parallel.experimental.iterators._iterators import IteratorBase
```
Comment (rwgk) [inline python/cuda_parallel/cuda/parallel/experimental/_cccl.py:205]: Will this work as a normal import near the top?

File: python/cuda_parallel/cuda/parallel/experimental/algorithms/reduce.py [hunk 3]
```diff
@@ -3,24 +3,26 @@
...
 from numba.cuda.cudadrv import enums
+from typing import Callable
 
 from .. import _cccl as cccl
 from .._bindings import get_paths, get_bindings
 
 
 class _Op:
-    def __init__(self, dtype, op):
+    def __init__(self, dtype: "np.dtype", op: Callable):
```
Comment (rwgk) [inline python/cuda_parallel/cuda/parallel/experimental/algorithms/reduce.py:18]: Oversight? (remove the double-quotes)

**Review by rwgk — APPROVED** [review]
LGTM for the mypy changes.

Did you mean to add in the thrust-related changes? (I looked through quickly, LGTM, but I don't have a lot of relevant background.)

**Comment by shwina** [comment]
> Did you mean to add in the thrust-related changes? (I looked through quickly, LGTM, but I don't have a lot of relevant background.)

Sorry for that - looks like a bad rebase, which I fixed. I'm going to go ahead and merge now. Thanks!

**Pull request merged**